### ISSUE TITLE
Add Meta-annotation Parameter Support

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-security#14480]
It would be nice if Spring Security's method security meta-annotation support allowed for parameters.

For example, it would be nice to be able to do:

```java
@Retention(RetentionPolicy.RUNTIME)
@PreAuthorize("hasAuthority(#value)")
public @interface HasAuthority {
    String value();
}
```

Then, an application could do:

```java
@HasAuthority("message:read")
public String method(...) {
}
```

The annotation expression should be able to handle method parameters, like `@PreAuthorize` already does:

```java
@Retention(RetentionPolicy.RUNTIME)
@PreAuthorize("@myAuthzBean.authorizeParameter(#object)")
public @interface AuthorizeObject {
}
```

allowing an application to do:

```java
@AuthorizeObject
public String method(Object object) {
}
```

If an annotation attribute and a method parameter have the same name, the method parameter value should take priority.

**Comment by kse-music** [comment]
@jzheaux  Can I do it?

**Comment by jzheaux** [comment]
Hi, @kse-music, thanks for the offer. I'm going to take care of this one for the moment since it is in a bit of flux (notice the recent changes in the description).

Please feel free to try out this branch to suggest improvements: https://github.com/jzheaux/spring-security/tree/gh-14480

**Comment by kse-music** [comment]
Oh, according to the previous description, I thought that the meta-annotation parameter was used as an SPEL expression variable, but I didn't expect it used as a placeholder parsed property sources. @jzheaux

**Comment by jzheaux** [comment]
That's right, @kse-music. As I was doing some of my own research, I found that approach to be easier. The reason is that SpEL does not support nested variables, which would be needed in some setups.

Sorry for the confusion. If you are still interested, I'd love to have your help on this PR: https://github.com/spring-projects/spring-security/pull/14494 once I have it out of draft.

**Comment by sbrannen** [comment]
Very interesting idea! 👍 

> The reason is that SpEL does not support nested variables, which would be needed in some setups.

Would you mind providing a concrete example just for the sake of clarity?

**Comment by jzheaux** [comment]
Sure, @sbrannen. In the following arrangement:

```java
@PreAuthorize("hasAuthority(#authority)")
public @interface HasAuthority {
    String authority();
}

// ...

@HasAuthority("#parameter")
public String method(String parameter) {
   // ...
}
```

what happens is if `method("value")` is invoked, the set of variables handed to the evaluation context looks something like:

```
{ "#parameter" : value, "#authority" : "#parameter" }
```

I couldn't quite unravel how to get SpEL to resolve `#authority` by first resolving `#parameter`.

**Comment by efenderbosch-atg** [comment]
Will this enhancement let me do something like this?

```kotlin
@PreAuthorize(
    """
    hasAuthority('PERM_FOO') and
    (#checkTenantId == false or #tenantId == authentication.principal.tenantId)
    """
)
annotation class HasFooPermission(val checkTenantId: Boolean = true)
```

```kotlin
@RestController
@SecurityRequirement(name = "Bearer Authentication")
@RequestMapping(value = "/foo", produces = [MediaType.APPLICATION_JSON_VALUE])
class FooController(private val service: FooService) {
    @HasFooPermission(checkTenantId = false)
    @GetMapping
    fun findAll(): ResponseEntity<List<Foo>> = service.findAll().ok()
}
```

**Comment by kse-music** [comment]
@efenderbosch-atg  According to the current implementation, attribute method variables in custom annotations need to be quoted through curly braces like so:

```
@PreAuthorize(
    """
    hasAuthority('PERM_FOO') and
    ({checkTenantId} == false or #tenantId == authentication.principal.tenantId)
    """
)
annotation class HasFooPermission(val checkTenantId: Boolean = true)
```
